docs/eip-draft_bls_remote_signer_api.md

```diff
@@ -0,0 +1,404 @@
+
+There are, however, use cases where it is required by the operator to run a validator client node in less constrained security environments, as the ones given by a cloud provider. Notwithstanding any security expectation, nothing prevents a rogue operator from gaining arbitrary access to the assets running inside a node.
+
+The situation is not better when the requirement is to execute the validators by leveraging a container orchestration solution (e.g. Kubernetes). The handling of secret keys across nodes can become a burden both from an operational as well as a security perspective.
+
+The proposed solution comprises running a specialized node with exclusive access to the secret keys, listening to a simple API (defined in the [Specification](#specification) section), and returning the requested signatures. Operators working under this schema must utilize clients with the adequate feature supporting the consumption of this API.
+
+The focus of this specification is the supply of BLS signatures _on demand_. The aspects of authentication, key management (creation, update, and deletion), pre-image validation, and transport encryption are discussed in the [Rationale](#rationale) section of this document. Moreover, the [Threat Model](#threat-model) section of this document provides a (non-exhaustive) list of threats and attack vectors, along with the suggested related mitigation strategy.
+
+## Specification
+
+### `GET /upcheck`
+
+_**Responses**_
+
+Success | <br>
+--- | ---
+Code | `200`
+Content | `{"status": "OK"}`
+
+---
+
+### `GET /publicKeys`
+
+_**Responses**_
+
+Success | <br>
+--- | ---
+Code | `200`
+Content | `{"public_keys": "[public_key_hex_string_without_0x]"}`
+
+_or_
+
+Error | <br>
+--- | ---
+Code | `404`
+Content | `{"error": "No keys found in storage."}`
+
+---
+
+### `POST /sign/:public_key`
+
+URL Parameter | <br>
+--- | ---
+`:public_key` | `public_key_hex_string_without_0x`
+
+_**Request**_
+
+JSON Body | <br> | <br>
+--- | --- | ---
+`signingRoot` | **Required** | A string representation of the hexadecimal value to be signed
+<br> | Optional | Any other field will be ignored by the signer
+
+_**Responses**_
+
+Success | <br>
+--- | ---
+Code |  `200`
+Content | `{"signature": "<signature_hex_string>"}`
+
+_or_
+
+Error | <br>
+--- | ---
+Code |  `400`
+Content | `{"error": "<Bad Request Error Message>"}`
+
+_or_
+
+Error | <br>
+--- | ---
+Code |  `404`
+Content | `{"error": "Key not found: <public_key_hex_string_without_0x>"}`
+
+---
+
+## Rationale
+
+### URL parameter `:public-key` without `0x`
```

Comment:
I don't see how the 0x-prefix prevents other id schemes.

The [validators endpoint](https://ethereum.github.io/eth2.0-APIs/#/Beacon/getStateValidators) in the standard API accepts either:

- Integer
- 0x-prefixed hex

Perhaps I'm missing something :)

Reply:
The idea here is that if one implementer wants to have their custom ids for their keys, for _whatever_ reason, we don't enforce in the API that they **must** have their id starting with `0x`.

Examples:

- `6adb11`: First 6 digits.
- `key-01`: Illegal hex characters

In particular, rust-bls-remote-signer takes the public key hex string without `0x`, and does a check looking for mismatches between the PK (given as param) and the retrieved SK.